run_id for SWPC

BryanTurns opened the pull request:
I modified the process_metadata function to now include a hash field in the metadata. It also sets the run_id for the SWPC institute to be the 5 digits in the data folder name while falling back to the hash for all other institutes.

greglucas commented on scripts/process_output.py:
```diff
@@ -307,29 +307,38 @@ def process_metadata(ds):
         The path of a new directory where all data should be stored for this run.
     """
     s = json.dumps(ds.attrs, cls=NumpyEncoder)
+    metadata = json.loads(s) 
```
Isn't this the same as `metadata=ds.attrs`?

BryanTurns replied:
I wanted to avoid any funkiness of directly editing the object within the DataSet Object

greglucas commented on scripts/process_output.py:
```diff
@@ -307,29 +307,38 @@ def process_metadata(ds):
+    # Get the run_id from the filename if the institute is SWPC, otherwise set it to the hash
+    if institute == "SWPC":
+        run_id = path.name.split("_")[-1].split(".")[0]
+        if not run_id:
+            raise ValueError("SWPC run metadata does not contain a run_id")
```
Is this possible to hit? I think your splits and indexing above would error before getting here.

greglucas commented on scripts/process_output.py:
```diff
@@ -307,29 +307,38 @@ def process_metadata(ds):
+    metadata = json.loads(s) 
     # Hash based on the metadata of the run
-    run_id = hashlib.sha256(s.encode("utf-8")).hexdigest()[:8]
+    hash = hashlib.sha256(s.encode("utf-8")).hexdigest()[:8]    
```
hash is a builtin function in Python, so we should probably call this something else to not override it and cause confusion. `hash_digest`?

greglucas reviewed (commented):
An idea to try and restructure this would be to try and get the run_id for a SWPC file immediately in a try/except block. Setting the run_id. Then later updating that with a hash if it isn't present.

```python
try:
    ds.attrs["run_id"] = path.name.split("_")[-1].split(".")[0]
except:
    ds.attrs["run_id"] = hashlib.sha256(s.encode("utf-8")).hexdigest()[:8]
```

BryanTurns commented:
> An idea to try and restructure this would be to try and get the run_id for a SWPC file immediately in a try/except block. Setting the run_id. Then later updating that with a hash if it isn't present.
> 
> ```python
> try:
>     ds.attrs["run_id"] = path.name.split("_")[-1].split(".")[0]
> except:
>     ds.attrs["run_id"] = hashlib.sha256(s.encode("utf-8")).hexdigest()[:8]
> ```

Just incase this doesn't end up throwing an exception for other institutes I am going to leave this in, but I'll add the suggested try except so that if the naming ever changes, there's a fall back.

BryanTurns commented:
Switched over to just storing data in the ds.attrs dictionary directly. I added the path to allow me to use this function in the AWS lambdas for parallel processing.

BryanTurns commented:
Actually we may want to change the way we process the run_id and instead use the bottom of the CONE files. The CDK code causes weirdness with using the path.

BryanTurns commented:
I verified that this works with the new CDK implementation for processing via lambda, the old style, and the manual way.

greglucas commented:
Looks like you have conflicts. Can you resolve those and push?

greglucas commented on scripts/process_output.py:
```diff
@@ -308,28 +309,40 @@ def process_metadata(ds):
+    # Extract the run_id from the dataset attributes
+    if not run_id:
+        # In order to extract the run_id from a SWPC run, the path must end with wsa_enlil_57484.57285344.dbqs0
+        if institute == "SWPC" and re.match(r"^.*wsa_enlil_\d{5}\.\d*\.dbqs0$", path.name):
```
Where does "path" come from here...? This seems like it is coming from outside of this scope, so is this a global or is there a way you can pass this path in instead of newpath?
